apps/gateway/test/champions_test.exs

```diff
@@ -0,0 +1,286 @@
+      assert_receive %WebSocketResponse{response_type: {:user, %User{}}}
+
+      fetch_last_message(socket_tester)
+      %WebSocketResponse{response_type: {:user, user}} = get_last_message()
+
+      # Creating another user with the same name fails
+      :ok = SocketTester.create_user(socket_tester, "Username")
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:error, %Error{reason: "username_taken"}}}
+
+      # Get user by name
+      :ok = SocketTester.get_user_by_username(socket_tester, "Username")
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:user, ^user}}
+    end
+  end
+
+  describe "units" do
+    test "unit selection", %{socket_tester: socket_tester} do
+      {:ok, user} = Users.register("SelectUser")
+
+      [unit_to_unselect | _] = user.units
+      slot = unit_to_unselect.slot
+
+      # Unit is selected by default (this will change when we remove sample data in user creation)
+      assert unit_to_unselect.selected
+
+      # Unselect the unit
+      :ok = SocketTester.unselect_unit(socket_tester, user.id, unit_to_unselect.id)
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:unit, %Unit{}}}
+
+      fetch_last_message(socket_tester)
+      %WebSocketResponse{response_type: {:unit, unselected_unit}} = get_last_message()
+
+      assert not unselected_unit.selected
+      # Protobuf doesn't support nil values, returns zero instead
+      assert unselected_unit.slot == 0
+
+      :ok = SocketTester.select_unit(socket_tester, user.id, unselected_unit.id, slot)
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:unit, %Unit{}}}
+
+      fetch_last_message(socket_tester)
+      %WebSocketResponse{response_type: {:unit, selected_unit}} = get_last_message()
+
+      assert selected_unit.selected
+      assert selected_unit.slot == slot
+    end
+
+    test "unit progression", %{socket_tester: socket_tester} do
+      muflus = GameBackend.Units.Characters.get_character_by_name("Muflus")
+      {:ok, user} = Users.register("LevelUpUser")
+      Currencies.add_currency_by_name!(user.id, "Gold", 9999)
+
+      {:ok, unit} =
+        GameBackend.Units.insert_unit(%{
+          user_id: user.id,
+          level: 19,
+          tier: 1,
+          rank: 2,
+          selected: false,
+          character_id: muflus.id
+        })
+
+      gold = Currencies.get_amount_of_currency_by_name(user.id, "Gold")
+      level = unit.level
+
+      #### Level up
+      [%CurrencyCost{currency_id: _gold_id, amount: level_up_cost}] = Units.calculate_level_up_cost(unit)
+
+      :ok = SocketTester.level_up_unit(socket_tester, user.id, unit.id)
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:unit_and_currencies, %UnitAndCurrencies{}}
+      }
+
+      fetch_last_message(socket_tester)
```

Comment:
I see that the call of `fetch_last_message` twice is very recurrent, is there a reason?

Reply:
Yes, the idea behind it was to separate between the send-receive behaviors and the validation of the data, but it resulted from an earlier integration of the SocketTester. I think there's no problem in doing both at the same time so I'm going to get rid of the extra ones!